components/file_system/src/rate_limiter.rs

```diff
@@ -1,74 +1,431 @@
+macro_rules! request_imp {
+    ($limiter:ident, $priority:ident, $amount:ident, $mode:tt) => {{
+        let priority_idx = $priority as usize;
+        let cached_bytes_per_epoch = $limiter.bytes_per_epoch[priority_idx].load(Ordering::Relaxed);
+        // Flow control is disabled when limit is zero.
+        if cached_bytes_per_epoch == 0 {
+            return $amount;
+        }
+        let mut amount = std::cmp::min($amount, cached_bytes_per_epoch);
+        let bytes_through =
+            $limiter.bytes_through[priority_idx].fetch_add(amount, Ordering::Relaxed) + amount;
+        // We prefer not to partially return only a portion of requested bytes.
+        if bytes_through <= cached_bytes_per_epoch
+            || !$limiter.strict && $priority == IOPriority::High
+        {
+            return amount;
         }
-        for i in self.write.iter() {
-            i.store(0, Ordering::Relaxed);
+        let mut wait = {
+            let now = Instant::now_coarse();
+            let mut locked = $limiter.protected.lock();
+            // The request is already partially fulfilled in current epoch when consumption
+            // overflow bytes are smaller than requested amount.
+            let remains = std::cmp::min(bytes_through - cached_bytes_per_epoch, amount);
+            // When there is a recent refill, double check if bytes consumption has been reset.
+            if now + DEFAULT_REFILL_PERIOD < locked.next_refill_time + Duration::from_millis(1)
+                && $limiter.bytes_through[priority_idx].fetch_add(remains, Ordering::Relaxed)
+                    + remains
+                    <= cached_bytes_per_epoch
+            {
+                return amount;
+            }
+            // Enqueue itself by adding to pending_bytes, whose current value denotes a position
+            // of logical queue to wait in.
+            locked.pending_bytes[priority_idx] += remains;
+            // Calculate wait duration by queue_len / served_per_epoch.
+            if locked.next_refill_time <= now {
+                $limiter.refill(&mut locked, now);
+                // Bytes served by next epoch (and skipped epochs) during refill are subtracted
+                // from pending_bytes, round up the rest.
+                DEFAULT_REFILL_PERIOD
+                    * ((locked.pending_bytes[priority_idx] + cached_bytes_per_epoch - 1)
+                        / cached_bytes_per_epoch) as u32
+            } else {
+                // `(a-1)/b` is equivalent to `roundup(a.saturating_sub(b)/b)`.
+                locked.next_refill_time - now
+                    + DEFAULT_REFILL_PERIOD
+                        * ((locked.pending_bytes[priority_idx] - 1) / cached_bytes_per_epoch) as u32
+            }
+        };
+        if wait > MAX_WAIT_DURATION_PER_REQUEST {
+            // Long wait duration could freeze request thread not to react to latest budgets
+            // adjustment. Exit early by returning partial quotas.
```

Comment:
what will the caller do when the write quota is partially satisified.

Reply:
We have retry logic in `File` and `rocksdb::Env` for this case.